src/morphology/german/stem.js

```diff
@@ -0,0 +1,178 @@
+	if ( c1Index !== -1 ) {
+		c1Index++;
+	}
+	let optionUsed1 = "";
+	let index1 = 10000;
+	if ( a1Index !== -1 ) {
+		optionUsed1 = "a";
+		index1 = a1Index;
+
+		return { index1, optionUsed1 };
+	} else if ( b1Index !== -1 ) {
+		optionUsed1 = "b";
+		index1 = b1Index;
+
+		return { index1, optionUsed1 };
+	} else if ( c1Index !== -1 ) {
+		optionUsed1 = "c";
+		index1 = c1Index;
+
+		return { index1, optionUsed1 };
+	}
+
+	return { index1, optionUsed1 };
+};
+
+/**
+ * Search for the longest among the following suffixes,
+ * (a) en   er   est
+ * (b) st (preceded by a valid st-ending, itself preceded by at least 3 letters)
+ * Define a valid st-ending as one of b, d, f, g, h, k, l, m, n or t.
+ *
+ * @param {string} word The word to check for the suffix.
+ * @returns {number} The index of the suffix.
+ */
+const findSuffixStep2 = function( word ) {
+	const a2Index = word.search( /(en|er|est)$/g );
+	let b2Index = word.search( /(.{3}[bdfghklmnt]st)$/g );
+	// Exclude the st-ending and the preceding 3 letters.
+	if ( b2Index !== -1 ) {
+		b2Index += 4;
+	}
+	let index2 = 10000;
+	if ( a2Index !== -1 ) {
+		index2 = a2Index;
+	} else if ( b2Index !== -1  ) {
+		index2 = b2Index;
+	}
+
+	return index2;
+};
+
+/**
+ * Delete the suffix found in step 1 if in R1. (The letter of the valid s-ending is not necessarily in R1.)
+ * If an ending of group (b) is deleted, and the ending is preceded by niss, delete the final s.
+ * (For example, äckern -> äck, ackers -> acker, armes -> arm, bedürfnissen -> bedürfnis).
+ *
+ * @param {string} word         The word for which to delete the suffix.
+ * @param {number} index1       The index of the suffix found in step 1.
+ * @param {string} optionUsed1  The type of the suffix found in step 1.
+ * @param {number} r1Index      The R1 index.
+ *
+ * @returns {string} The word with the deleted suffix.
+ */
+const deleteSuffix1 = function( word, index1, optionUsed1, r1Index ) {
+	if ( index1 !== 10000 && r1Index !== -1 ) {
+		if ( index1 >= r1Index ) {
+			word = word.substring( 0, index1 );
+			if ( optionUsed1 === "b" ) {
+				if ( word.search( /niss$/ ) !== -1 ) {
+					word = word.substring( 0, word.length - 1 );
+				}
+			}
+		}
+	}
+	return word;
+};
+
+/**
+ * Delete the suffix found in step 2 if in R1.
+ * (For example, derbsten -> derbst by step 1, and derbst -> derb by step 2,
+ * since b is a valid st-ending, and is preceded by just 3 letters).
+ *
+ * @param {string} word     The word for which to delete the suffix.
+ * @param {number} index2   The index of the suffix found in step 2.
+ * @param {number} r1Index  The R1 index.
+ *
+ * @returns {string} The word with the deleted suffix.
+ */
+const deleteSuffix2 = function( word, index2, r1Index ) {
+	if ( index2 !== 10000 && r1Index !== -1 ) {
+		if ( index2 >= r1Index ) {
+			word = word.substring( 0, index2 );
+		}
+	}
+	return word;
+};
+
+/**
+ * Stems German words.
+ *
+ * Stem function adapted from:
+ * Author: Joder Illi
```

Comment:
Perhaps we can markup this JSDoc comment with some fancy tags like [`@author`](http://usejsdoc.org/tags-author.html) and [`@copyright`](http://usejsdoc.org/tags-copyright.html) 🙂.
Although I am not sure whether `@copyright` works in this context.